Installation issue: mxnet

<!-- Thanks for taking the time to report this build failure. To proceed with the report please:

1. Title the issue "Installation issue: <name-of-the-package>".
2. Provide the information required below.

We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively! -->

Building mxnet with CUDA fails

### Steps to reproduce the issue

<!-- Fill in the exact spec you are trying to build and the relevant part of the error message -->
```console
$ spack install mxnet +cuda
==> Error: ProcessError: Command exited with status 2:
    'make' 'CC=/usr/bin/gcc' 'CXX=/usr/bin/g++' 'MSHADOW_PATH=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/mshadow-20170721-6g2snd4q2odnk6dzeedto3sqmk6jfybc' 'PS_PATH=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/ps-lite-20170328-m7nltcy7jnxiphhizkgs5hkzp4wuxpmd' 'USE_OPENMP=0' 'USE_CUDA=1' 'USE_CUDNN=1' 'USE_OPENCV=1' 'USE_PROFILER=0' 'USE_OPENCV_INC_PATH=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/opencv-4.5.0-7jpy5ia3smym6j3ncls7wlzspkkskc4g/include' 'USE_OPENCV_LIB_PATH=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/opencv-4.5.0-7jpy5ia3smym6j3ncls7wlzspkkskc4g/lib' 'USE_BLAS=openblas' 'USE_CUDA_PATH=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/cuda-11.1.1-lht3e45eusx255yymqezp6cchlmranl5' 'CUDNN_PATH=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/cudnn-8.0.4.30-11.1-j4abrfxwhhugumn3xv3nh4mm4huvxxda' 'CUB_INCLUDE=/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/cub-1.9.10-ix3blep6ns7hllhgpuybhi4mmof6of6s/include'

1 error found in build log:
     21    /usr/bin/g++ -std=c++11 -c -DMSHADOW_FORCE_STREAM -Wall -Wsign-compare -DDMLC_MODERN_THREAD_LOCAL=0 -O3 -DNDEBUG=1 -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/mshadow
           / -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/dmlc-core/include -fPIC -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3
           rdparty/tvm/nnvm/include -I/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/openblas-0.3.13-72iaujajp5u3fmd6wifbjvykphj4hl2d/include -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckj
           irsnkqcenwcal/spack-src/3rdparty/dlpack/include -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/tvm/include -Iinclude -funroll-loops -Wno-unused-parameter -Wno-unknown-pr
           agmas -Wno-unused-local-typedefs -msse3 -mf16c -I/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/cuda-11.1.1-lht3e45eusx255yymqezp6cchlmranl5/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_R
           ABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMSHADOW_USE_PASCAL=0 -DMXNET_USE_OPENCV=1 -I/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/opencv-4.5.0-7jpy5ia3smym6j3ncls7wlzspkkskc4g/include/opencv4 -DMXNET_USE_OPERA
           TOR_TUNING=1 -DMSHADOW_INT64_TENSOR_SIZE=0 -DMSHADOW_USE_CUDNN=1 -DMXNET_USE_BLAS_OPEN=1  -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/nvidia_cub -DMXNET_ENABLE_CUDA_R
           TC=1 -DMXNET_USE_NCCL=0 -DMXNET_USE_LIBJPEG_TURBO=0 -MMD -c src/operator/subgraph/tensorrt/nnvm_to_onnx.cc -o build/src/operator/subgraph/tensorrt/nnvm_to_onnx.o
     22    /usr/bin/g++ -std=c++11 -c -DMSHADOW_FORCE_STREAM -Wall -Wsign-compare -DDMLC_MODERN_THREAD_LOCAL=0 -O3 -DNDEBUG=1 -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/mshadow
           / -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/dmlc-core/include -fPIC -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3
           rdparty/tvm/nnvm/include -I/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/openblas-0.3.13-72iaujajp5u3fmd6wifbjvykphj4hl2d/include -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckj
           irsnkqcenwcal/spack-src/3rdparty/dlpack/include -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/tvm/include -Iinclude -funroll-loops -Wno-unused-parameter -Wno-unknown-pr
           agmas -Wno-unused-local-typedefs -msse3 -mf16c -I/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/cuda-11.1.1-lht3e45eusx255yymqezp6cchlmranl5/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_R
           ABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMSHADOW_USE_PASCAL=0 -DMXNET_USE_OPENCV=1 -I/home/bjonglez/spack/opt/spack/linux-debian10-skylake_avx512/gcc-8.3.0/opencv-4.5.0-7jpy5ia3smym6j3ncls7wlzspkkskc4g/include/opencv4 -DMXNET_USE_OPERA
           TOR_TUNING=1 -DMSHADOW_INT64_TENSOR_SIZE=0 -DMSHADOW_USE_CUDNN=1 -DMXNET_USE_BLAS_OPEN=1  -I/tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/nvidia_cub -DMXNET_ENABLE_CUDA_R
           TC=1 -DMXNET_USE_NCCL=0 -DMXNET_USE_LIBJPEG_TURBO=0 -MMD -c src/operator/numpy/linalg/np_gesvd.cc -o build/src/operator/numpy/linalg/np_gesvd.o
     23    In file included from /tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/mshadow/mshadow/tensor.h:16,
     24                     from include/mxnet/./base.h:33,
     25                     from include/mxnet/operator_util.h:43,
     26                     from src/operator/numpy/linalg/np_gesvd.cc:25:
  >> 27    /tmp/bjonglez/spack-stage/spack-stage-mxnet-1.6.0-myghj6o6rfetibmyckjirsnkqcenwcal/spack-src/3rdparty/mshadow/mshadow/./base.h:179:12: fatal error: cudnn.h: No such file or directory
     28       #include <cudnn.h>
     29                ^~~~~~~~~
     30    compilation terminated.
     31    make: *** [Makefile:572: build/src/operator/numpy/linalg/np_gesvd.o] Error 1
```

### Information on your system

<!-- Please include the output of `spack debug report` -->
* **Spack:** 0.16.0-950-bfa480d87d
* **Python:** 3.7.3
* **Platform:** linux-debian10-skylake_avx512
* **Concretizer:** original

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

### Additional information

<!-- Please upload the following files. They should be present in the stage directory of the failing build. Also upload any config.log or similar file if one exists. -->
* [spack-build-out.txt](https://github.com/spack/spack/files/5868153/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/5868157/spack-build-env.txt)

<!-- Some packages have maintainers who have volunteered to debug build failures. Run `spack maintainers <name-of-the-package>` and @mention them here if they exist. -->
@adamjstewart

### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/mxnet/makefile-cuda-stub-1.6.patch
+++ b/var/spack/repos/builtin/packages/mxnet/makefile-cuda-stub-1.6.patch
@@ -1,0 +1,13 @@
+diff --git a/Makefile b/Makefile
+index 4746cc434..335c79b64 100644
+--- a/Makefile
++++ b/Makefile
+@@ -518,7 +518,7 @@ ifeq ($(USE_CUDA), 1)
+ 	endif
+ 	# Make sure to add stubs as fallback in order to be able to build
+ 	# without full CUDA install (especially if run without nvidia-docker)
+-	LDFLAGS += -L/usr/local/cuda/lib64/stubs
++	LDFLAGS += -L$(USE_CUDA_PATH)/lib64/stubs
+ 	ifeq ($(USE_NCCL), 1)
+ 		ifneq ($(USE_NCCL_PATH), NONE)
+ 			CFLAGS += -I$(USE_NCCL_PATH)/include

--- a/var/spack/repos/builtin/packages/mxnet/makefile-cudnn-path-1.6.patch
+++ b/var/spack/repos/builtin/packages/mxnet/makefile-cudnn-path-1.6.patch
@@ -1,0 +1,13 @@
+diff --git a/Makefile b/Makefile
+index 4746cc434..e738fc53f 100644
+--- a/Makefile
++++ b/Makefile
+@@ -509,7 +509,7 @@ LIB_DEP += $(DMLC_CORE)/libdmlc.a $(NNVM_PATH)/lib/libnnvm.a
+ ALL_DEP = $(OBJ) $(EXTRA_OBJ) $(PLUGIN_OBJ) $(LIB_DEP)
+ 
+ ifeq ($(USE_CUDA), 1)
+-	CFLAGS += -I$(ROOTDIR)/3rdparty/nvidia_cub
++	CFLAGS += -I$(CUB_INCLUDE) -I$(CUDNN_PATH)/include
+ 	ALL_DEP += $(CUOBJ) $(EXTRA_CUOBJ) $(PLUGIN_CUOBJ)
+ 	LDFLAGS += -lcufft
+ 	ifeq ($(ENABLE_CUDA_RTC), 1)

--- a/var/spack/repos/builtin/packages/mxnet/package.py
+++ b/var/spack/repos/builtin/packages/mxnet/package.py
@@ -44,8 +44,9 @@ class Mxnet(MakefilePackage):
     depends_on('py-numpy@1.16:', type=('build', 'run'), when='@1.6.0 +python')
     extends('python', when='+python')
 
-    patch('makefile.patch', when='@0.10:0.11')
     patch('makefile.opencv.patch', when='@1.6.0')
+    patch('makefile-cudnn-path-1.6.patch', when='@1.6.0')
+    patch('makefile-cuda-stub-1.6.patch', when='@1.6.0')
     patch('parallell_shuffle.patch', when='@1.6.0')
 
     def build(self, spec, prefix):

